File: 4104-shortest-path-with-at-most-k-consecutive-identical-characters/shortest-path-with-at-most-k-consecutive-identical-characters.cpp
struct compare {
    bool operator()(vector<int>& a, vector<int>& b) {
        return a[2] > b[2];
    }
};

class Solution {
public:
    int shortestPath(int n, vector<vector<int>>& edges, string labels, int k) {
        vector<vector<vector<int>>> adj(n);
        for(vector<int>& edge: edges) {
            adj[edge[0]].push_back({edge[1], edge[2]});
        }

        priority_queue<vector<int>, vector<vector<int>>, compare> pq;
        vector<vector<bool>> visited(n, vector<bool>(k+1, false));

        pq.push({0, 1, 0});

        while(!pq.empty()) {
            auto rv = pq.top();
            pq.pop();
            if(visited[rv[0]][rv[1]]) {
                continue;
            }
            if(rv[0] == n - 1) {
                return rv[2];
            }
            visited[rv[0]][rv[1]] = true;
            for(vector<int>& nbrs: adj[rv[0]]) {
                int newWeight = rv[2] + nbrs[1];
                int cnt = (labels[rv[0]]-'a' == (labels[nbrs[0]]-'a')) ? rv[1]+1 : 1;
                if(cnt > k || visited[nbrs[0]][cnt]) {
                    continue;
                }
                pq.push({nbrs[0], cnt, newWeight});
            }
        }

        return -1;
    }
};

File: 4104-shortest-path-with-at-most-k-consecutive-identical-characters/shortest-path-with-at-most-k-consecutive-identical-characters.java
class Solution {
    public int shortestPath(int n, int[][] edges, String labels, int k) {
        List<int[]>[] adj = new ArrayList[n];
        
        for(int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for(int[] edge: edges) {
            adj[edge[0]].add(new int[]{edge[1], edge[2]});
        }

        int[][] dist = new int[n][k+1];
        for(int[] d: dist) {
            Arrays.fill(d, Integer.MAX_VALUE);
        }
        dist[0][1] = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[2], b[2]));
        pq.add(new int[]{0, 1, 0});

        while(!pq.isEmpty()) {
            int[] rv = pq.poll();

            if(dist[rv[0]][rv[1]] < rv[2]) {
                continue;
            }

            if(rv[0] == n - 1) {
                return rv[2];
            }

            for(int[] nbrs: adj[rv[0]]) {
                int newD = dist[rv[0]][rv[1]] + nbrs[1];
                int c = (labels.charAt(rv[0]) == labels.charAt(nbrs[0])) ? 1+rv[1] : 1;
                if(c > k || dist[nbrs[0]][c] <= newD) {
                    continue;
                }
                dist[nbrs[0]][c] = newD;
                pq.add(new int[]{nbrs[0], c, newD});
            }
        }

        return -1;
    }
}